Sprint 17

AlfaBotan opened the pull request:
Я не знаю как я это сделал, но я это сделал!
Сильно не ругайтесь, если заметите какие-то опечатки или что-то подобное, я 28 часов подряд просидел за проектом что бы вовремя его сдать)
по требованиям перед сдачей пробежался пару раз, и если я все их правильно понял, то весь необходимый функционал должен присутствовать.
Буду рад если оставите подробные комментарии к рекомендациям, ну а если вдруг будут критические ошибки то пожалуйста, растолкуйте что там к чему, не решение готовое конечно, но что бы можно было сразу исправить😊
Заранее спасибо!

AlfaBotan committed:
Добавлена локализация главного экрана для русского и английского языков

AlfaBotan committed:
добавил тесты для светлой темы

настроил правильное отображение элементов при включении темный темы

AlfaBotan committed:
Добавлен функционал сбора аналитических данных (пока не для всех требуемых ивентов)
Добавлен функционал поиска трекеров, на выбранную дату

AlfaBotan committed:
Добавлен функционал контекстного меню для ячеек коллекции:
Кнопки : Закрепить\Открепить, редактировать, удалить.
Реализованы соответствующие методы, так же создан экран для редактирования

AlfaBotan committed:
реализован функционалпоиска фильтров, а так же сортировки

AlfaBotan committed:
Реализован экран статистики

andrew-kuzn-yndx commented on Tracker/CategoryViewController/View/CategoryTableViewCell.swift:
```diff
@@ -73,6 +74,7 @@ final class CategoryTableViewСеll: UITableViewCell {
         }
     }
     
+    
```
Давай уберем лишнюю строку, будет симпатичнее разделять блоки кода одной пустой строкой)

andrew-kuzn-yndx commented on Tracker/CoreData/CoreDataManager.swift:
```diff
@@ -60,12 +59,49 @@ final class CoreDataManager: NSObject {
+        // Задаем сортировку (если нужно, можно изменить ключ сортировки)
+        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "category.title", ascending: true)]
+        
+        // Настраиваем или обновляем fetchedResultsController
```
Лучше удалять такие избыточные комментарии в коде. Я понимаю, что это нужно при работе над проекте, но правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код)
Естественно это не касается важных комментариев

AlfaBotan replied:
Добрый вечер, прошелся я по рекомендациям, и критическим замечаниям, готов в целом отправлять, хотел бы у вас только у точнить 1 момент, вы писали в критических замечаниях, что фильтр не сбрасывался даже при перезапуске приложения, я вот немного не пойму как такое у вас могло произойти, если у меня в AppDelegate определены методы, для очистки фильтра в UserDefault:

func applicationWillTerminate(_ application: UIApplication) {
            UserDefaults.standard.removeObject(forKey: "pickedFilter")
        }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
            UserDefaults.standard.removeObject(forKey: "pickedFilter")
        }.
        
ну и я на всякий случай ручками потестировал, он действительно сбрасывается.
Хотя на данном этапе реализации приложения это не так важно, фильтр теперь не пропадает навсегда без возможности его снова увидеть, но вероятно это был какой-то баг который срабатывает у вас, и я хотел бы его подправить)

andrew-kuzn-yndx commented on Tracker/FiltersViewController/FiltersTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+        layer.masksToBounds = false
+    }
+    
+    func configureSubviews() {
```
Стоит добавить `private`
Необходимо указывать private  для того, чтобы спрятать детали реализации конкретной части функциональности, 
когда они используются только внутри области объявления.

andrew-kuzn-yndx commented on Tracker/FiltersViewController/FiltersTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+            doneImg.isHidden = false
+        } else {
+            doneImg.isHidden = true
+        }
```
Тут можно упростить через "инвертор"
doneImg.isHidden = !doneImg.isHidden

andrew-kuzn-yndx commented on Tracker/FiltersViewController/FiltersViewController.swift:
```diff
@@ -0,0 +1,135 @@
+protocol FilterViewControllerProtocol: AnyObject {
+    func saveChoise(filter: FiltersCases)
+}
+final class FiltersViewController: UIViewController {
```
А тут наоборот, можно добавить пустых строк, чтобы немного отделить блоки друг от друга)

andrew-kuzn-yndx commented on Tracker/FiltersViewController/FiltersViewController.swift:
```diff
@@ -0,0 +1,135 @@
+        titleLable.text = "Фильтры"
+        titleLable.font = .systemFont(ofSize: 16, weight: .medium)
+        titleLable.textAlignment = .center
+        titleLable.textColor = .ypBlack
```
Для создания и настройки элемента, ты можешь воспользоваться следующей конструкцией, которая помогает настраивать элемент при его объявлении и не создавать дополнительные методы по настройке:
    private lazy var titleLable: UILabel = {
        let label = UILabel()
        label.textAlignment = .center
         ...реализация и настройка
        return label
    }()

andrew-kuzn-yndx commented on Tracker/FiltersViewController/FiltersViewController.swift:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+extension FiltersViewController: UITableViewDelegate {
```
Молодец, что вывел реализацию делегата в отдельный extension!

andrew-kuzn-yndx commented on Tracker/StatisticViewController/StatisticViewController.swift:
```diff
@@ -7,11 +7,135 @@
+        placeholder.image = UIImage(named: "statisticNil")
+        placeholderLable.font = .systemFont(ofSize: 12, weight: .medium)
+        placeholderLable.text = "Анализировать пока нечего"
+        placeholderLable.textAlignment = .center
```
Стоит пройтись по проекту и использовать 'немедленную инициализацию' вьюшек через lazy var, как я подсказывал тебе выше
Так будет проще читать код, это улучшит производительность, да и просто симпатичнее станет)
Вот тут в середине статьи есть примеры: https://dzen.ru/a/YmMsexP4yGGM3uf1

andrew-kuzn-yndx commented on Tracker/TrackerViewController/TrackerCollectionViewCell.swift:
```diff
@@ -103,9 +117,19 @@ final class TrackerCollectionViewCell: UICollectionViewCell {
     }
     
-    func configCell(id: UUID, name: String, color: UIColor, emoji: String, completedDays: Int, isEnabled: Bool, isCompleted: Bool, indexPath: IndexPath){
+    func configCell(id: UUID,
```
Можно использовать комментарии // MARK: - ... для организации кода.
Подробнее можно посмотреть в [этой статье](https://habr.com/ru/articles/744470/): https://habr.com/ru/articles/744470/

Тут, например, можно разделить методы на приватные и публичные блоки

andrew-kuzn-yndx commented:
Привет!)
Насчет 28 часов - очень понимаю 😄 Бывает так увлечешься, что сложно отвлечься)
Такие старания - хороший знак, ты молодец!

andrew-kuzn-yndx commented on Tracker/StatisticViewController/StatisticViewController.swift:
```diff
@@ -7,11 +7,135 @@
+        } else {
+            holderForStatistic.isHidden = false
+            placeholder.isHidden = true
+            placeholderLable.isHidden = true
```
тут можно сократить количество кода вдвое:

let isEmpty = countDays == 0

holderForStatistic.isHidden = isEmpty
placeholder.isHidden = !isEmpty
placeholderLable.isHidden = !isEmpty

AlfaBotan commented:
> Привет!) Насчет 28 часов - очень понимаю 😄 Бывает так увлечешься, что сложно отвлечься) Такие старания - хороший знак, ты молодец!
я вот не вспомню что бы вы у меня раньше проверяли, но такое объёмное ревью я получил впервые, да еще и со скриншотами и видео в диске, это вообще огонь!
Хотел бы уточнить по ревью)
Рекомендации, это понятно, я там чуть причешу код, вот разве что моментальная инициализация свойств UI объект мне как-то совсем не по душе, но я почитаю статью, если это влияет на производительность в лучшую сторону, то и это переделаю)
Но это так лирическое отступление)
У меня пара вопросов по критическим:

1) отмечать трекеры я целенаправленно разрешил, только день в день, подумал что так будет правильно, ибо отмечать ранее невыполнение трекеры как будто бы неверно, но если был такой пункт в требованиях, а я пропустил, то без проблем, я это исправлю, запретив только в будущем отмечать)

2) по фильтру "на сегодняшний день" там же прямо в ТЗ указано, что разница между фильтрами Все трекеры и На сегодняшний день, в том и заключается, что трекер на сегодняшний день перебрасывает нас "На сегодня" 
я тогда не совсем понимаю что делает фильтр "На сегодняшний день" можете немного прояснить ситуацию, а то я не совсем понимаю что исправлять в данном случае)

3) не совсем понимаю почему вам пришлось удалять приложение, ведь состояние фильтра сбрасывается после закрытия приложения у меня в appDelegate определены для этого методы:

    func applicationWillTerminate(_ application: UIApplication) {
            // Очистка состояния фильтра из UserDefaults
            UserDefaults.standard.removeObject(forKey: "pickedFilter")
        }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
            // Очистка состояния фильтра из UserDefaults
            UserDefaults.standard.removeObject(forKey: "pickedFilter")
        }
 ну и я вот сейчас перепроверил, фильтр сбрасывается)
 
4) по поводу выделения, оно же сбрасывает сразу же как только пользователь отпустил палец, ну то есть там же настроено это, что выделение не остается на всегда, это разве плохо?

Вы только не подумайте, что я пытаюсь оспорить ваше решение)
Я просто хотел бы понять где я ошибся, что бы сильно не затягивать с правками!

AlfaBotan committed:
причесал код, изменил логику скрытия кнопки "фильтр"

AlfaBotan committed:
изменил логику срабатывания алерта при выполнении трекера, теперь алерт срабатывает только "В будущем"

AlfaBotan committed:
Убрал выделение ячеек в таблицах
Изменил логику действий при выбранном фильтре "На сегодня" теперь при смене даты, устанавливается новый фильтр "Все трекеры"

AlfaBotan committed:
Добавлена локализация для кнопки Фильтры

AlfaBotan committed:
Изменил верстку в ячейках создания и редактирования трекера, теперь если категория или расписание не были выбраны, то название ячейки распологается в центре.

AlfaBotan committed:
Нашел баг, со странным поведением после нажатия на + в ячейке трекера, ИСПРАВЛЕНО

AlfaBotan committed:
Изменил таргет для тестов на 13.4
